File: Android/app/src/main/java/com/pokeapi/lpiem/pokeapiandroid/Provider/AppProviderSingleton.java
package com.pokeapi.lpiem.pokeapiandroid.Provider;

import com.pokeapi.lpiem.pokeapiandroid.Provider.SocialNetworks.FacebookApiProvider;
import com.pokeapi.lpiem.pokeapiandroid.Provider.SocialNetworks.GoogleApiProvider;
import com.pokeapi.lpiem.pokeapiandroid.Provider.SocialNetworks.TwitterApiProvider;

public class AppProviderSingleton {
    public final static int FACEBOOK = 1;
    public final static int TWITTER = 2;
    public final static int GOOGLE = 3;
    public final static int POKEAPI = 4;

    private static AppProviderSingleton instance;
    private FacebookApiProvider facebookApiProvider;
    private GoogleApiProvider googleApiProvider;
    private TwitterApiProvider twitterApiProvider;

    public static AppProviderSingleton getInstance(){
        if(instance == null){
            instance = new AppProviderSingleton();
        }
        return instance;
    }

    private AppProviderSingleton() {
        facebookApiProvider = new FacebookApiProvider();
    }

    public FacebookApiProvider getFacebookApiProvider() {
        return facebookApiProvider;
    }

    public void setFacebookApiProvider(FacebookApiProvider facebookApiProvider) {
        this.facebookApiProvider = facebookApiProvider;
    }

    public GoogleApiProvider getGoogleApiProvider() {
        return googleApiProvider;
    }

    public void setGoogleApiProvider(GoogleApiProvider googleApiProvider) {
        this.googleApiProvider = googleApiProvider;
    }

    public TwitterApiProvider getTwitterApiProvider() {
        return twitterApiProvider;
    }

    public void setTwitterApiProvider(TwitterApiProvider twitterApiProvider) {
        this.twitterApiProvider = twitterApiProvider;
    }
}

File: Android/app/src/main/java/com/pokeapi/lpiem/pokeapiandroid/Provider/AppProviderSingleton.kt
package com.pokeapi.lpiem.pokeapiandroid.Provider

import android.util.Log
import com.pokeapi.lpiem.pokeapiandroid.View.Fragment.PokedexListView
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import com.google.firebase.auth.FirebaseUser
import com.pokeapi.lpiem.pokeapiandroid.Model.Retrofit.Pokemons.PokemonList
import com.pokeapi.lpiem.pokeapiandroid.R


object AppProviderSingleton {
    private lateinit var firebaseUser:FirebaseUser
    var User:FirebaseUser
        get() = firebaseUser
        set(newValue){
            firebaseUser = newValue
        }

    /**
     * Fetch data from webservice
     * @param pokedexView fragment support for the data to be loaded
     */
    fun getPokeList(pokedexView: PokedexListView){
        RetrofitSingleton.buildInstance()!!.getPokemonListData().enqueue(object : Callback<PokemonList> {

            override fun onResponse(call: Call<PokemonList>, response: Response<PokemonList>) {
                if (response.isSuccessful) {
                    val returnedData = response.body()
                    pokedexView.initPokedex(returnedData!!.pokemonList)
                } else {
                    Log.d(pokedexView.context!!.getString(R.string.error_tag), pokedexView.context!!.getString(R.string.fetching_data_error))
                }
            }

            override fun onFailure(call: Call<PokemonList>, t: Throwable) {
                Log.e(pokedexView.context!!.getString(R.string.error_tag), t.localizedMessage)
                t.printStackTrace()
            }
        })
    }
}
